packages/story-editor/src/app/pageDataUrls/pageDataUrlsProvider.js

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+import { useMemo, useRef, useCallback, useState } from '@web-stories-wp/react';
+
+/**
+ * Internal dependencies
+ */
+import {
+  requestIdleCallback,
+  cancelIdleCallback,
+} from '../../utils/idleCallback';
+import storyPageToDataUrl from '../../utils/storyPageToDataUrl';
+import Context from './context';
+
+/**
+ * @typedef {import('../../types.js').Page} Page
+ */
+
+function PageDataUrlProvider({ children }) {
```

Comment:
Good idea with the task queue 👍  I might steal this for the media library. Maybe we could turn it into a reusable `useTaskQueue()` hook or something.

Reply:
yea I think it would be relatively straightforward to extract out and just plug in play in whatever providers we want to utilize it in.

Happy to do that here, but maybe nicer to do in a separate PR just this one's not so bloated (still need to add a fair amount of test coverage to this and tidy up some weird behavior)